packages/arb-token-bridge-ui/src/pages/api/cctp/[type].ts

```diff
@@ -0,0 +1,225 @@
+      })
+      return
+    }
+
+    const l1Subgraph = getSubgraphClient(
+      l1ChainId === ChainId.Mainnet ? 'cctp-mainnet' : 'cctp-goerli'
+    )
+    const l2Subgraph = getSubgraphClient(
+      l1ChainId === ChainId.Mainnet ? 'cctp-arb-one' : 'cctp-arb-goerli'
+    )
+
+    const messageSentsQuery = gql(`{
+        messageSents(
+          where: {
+            sender: "${walletAddress}"
+          }
+          orderDirection: "desc"
+          orderBy: "blockTimestamp"
+        ) {
+          attestationHash
+          blockNumber
+          blockTimestamp
+          id
+          message
+          nonce
+          sender
+          sourceDomain
+          transactionHash
+        }
+      }`)
+
+    const messageReceivedsQuery = gql(`{
+        messageReceiveds(
+          where: {
+            caller: "${walletAddress}"
+          }
+          orderDirection: "desc"
+          orderBy: "blockTimestamp"
+        ) {
+          id
+          caller
+          sourceDomain
+          nonce
+          blockTimestamp
+          blockNumber
+          messageBody
+          sender
+          transactionHash
+        }
+      }
+    `)
+
+    let messageSentsResult: ApolloQueryResult<{ messageSents: MessageSent[] }>
+    let messageReceivedsResult: ApolloQueryResult<{
+      messageReceiveds: MessageReceived[]
+    }>
+    if (type === 'deposits') {
+      ;[messageSentsResult, messageReceivedsResult] = await Promise.all([
+        l1Subgraph.query({ query: messageSentsQuery }),
+        l2Subgraph.query({ query: messageReceivedsQuery })
+      ])
+    } else {
+      ;[messageSentsResult, messageReceivedsResult] = await Promise.all([
+        l2Subgraph.query({ query: messageSentsQuery }),
+        l1Subgraph.query({ query: messageReceivedsQuery })
+      ])
+    }
+
+    const { messageSents } = messageSentsResult.data
+    const { messageReceiveds } = messageReceivedsResult.data
+
+    // MessageSents can be link to MessageReceived with the tuple (sourceDomain, nonce)
+    // Map constructor accept an array of [key, value] arrays
+    // new Map(['key1', 'value1'], ['key2', 'value2'], ['keyN', 'valueN']) would return
+    // Map(3) {'key1' => 'value1', 'key2' => 'value2', 'keyN' => 'valueN'}
+    // We create a map with all keys being MessageReceived ids, and values being the corresponding MessageReceived
+    const messageReceivedMap = new Map(
+      messageReceiveds.map(messageReceived => [
+        messageReceived.id,
+        messageReceived
+      ])
+    )
+    const { pending, completed } = messageSents.reduce(
+      (acc, message) => {
+        // If the MessageSent has a corresponding MessageReceived
+        if (messageReceivedMap.has(message.id)) {
+          acc.completed.push(message)
+        } else {
+          acc.pending.push(message)
+        }
+
+        return acc
+      },
+      {
+        completed: [],
+        pending: []
+      } as {
+        completed: MessageSent[]
+        pending: MessageSent[]
```

Comment:
For completed deposits/withdrawals, we'll probably want to display transactions from both chains (send and receive), so maybe the response should look more like this:

```
export type PendingCCTPTransfer = {
  messageSent: MessageSent
}

export type CompletedCCTPTransfer = PendingCCTPTransfer & {
  messageReceived: MessageReceived
}

export type Response =
  | {
      data: {
        pending: PendingCCTPTransfer[]
        completed: CompletedCCTPTransfer[]
      }
      error: null
    }
  | {
      data: null
      error: string
    }
```